packages/xl-ai/src/AIExtension.ts

```diff
@@ -0,0 +1,293 @@
+            }
+          }
+          return true;
+        },
+      }),
+    );
+    this.addProsemirrorPlugin(suggestChanges());
+    this.addProsemirrorPlugin(createAgentCursorPlugin(options.agentCursor));
+  }
+
+  /**
+   * Open the AI menu at a specific block
+   */
+  public openAIMenuAtBlock(blockID: string) {
+    this.editor.setForceSelectionVisible(true);
+    this.editor.isEditable = false;
+    this._store.setState({
+      aiMenuState: {
+        blockId: blockID,
+        status: "user-input",
+      },
+    });
+  }
+
+  /**
+   * Close the AI menu
+   */
+  public closeAIMenu() {
+    this._store.setState({
+      aiMenuState: "closed",
+    });
+    this.editor.setForceSelectionVisible(false);
+    this.editor.isEditable = true;
+    this.editor.focus();
+  }
+
+  /**
+   * Accept the changes made by the LLM
+   */
+  public acceptChanges() {
+    this.editor.exec(applySuggestions);
+    this.closeAIMenu();
+  }
+
+  /**
+   * Reject the changes made by the LLM
+   */
+  public rejectChanges() {
+    this.editor.exec(revertSuggestions);
+    this.closeAIMenu();
+  }
+
+  /**
+   * Update the status of a call to an LLM
+   *
+   * @warning This method should usually only be used for advanced use-cases
+   * if you want to implement how an LLM call is executed. Usually, you should
+   * use {@link callLLM} instead.
+   */
+  public setAIResponseStatus(
+    status:
+      | "user-input"
+      | "thinking"
+      | "ai-writing"
+      | "error"
+      | "user-reviewing",
+  ) {
+    const aiMenuState = this.store.getState().aiMenuState;
+    if (aiMenuState === "closed") {
+      return; // TODO: log error?
+    }
+
+    if (status === "ai-writing") {
+      this.editor.setForceSelectionVisible(false);
+    }
+
+    this._store.setState({
+      aiMenuState: {
+        status: status,
+        blockId: aiMenuState.blockId,
+      },
+    });
+  }
+
+  /**
+   * Execute a call to an LLM and apply the result to the editor
+   */
+  public async callLLM(opts: CallSpecificCallLLMOptions) {
+    const options = {
+      ...this.options.getState(),
+      ...opts,
+    };
+
+    this.setAIResponseStatus("thinking");
+
+    try {
+      let ret: CallLLMResult;
+
+      // TODO: maybe separate functions for streaming / non-streaming?
+      if (options.dataFormat === "json") {
+        throw new Error("not implemented");
+        // ret = await llm.json.call(this.editor, options);
+      } else if (options.dataFormat === "markdown") {
+        throw new Error("not implemented");
+        // ret = await llm.markdown.call(this.editor, options);
+      } else if (options.dataFormat === "html") {
+        ret = await llm.html.call(this.editor, {
+          ...options,
+          onStart: () => {
+            this.setAIResponseStatus("ai-writing"); // TODO: also apply to other formats
+          },
+          onBlockUpdate: (blockId: string) => {
+            // NOTE: does this setState with an anon object trigger unnecessary re-renders?
```

Comment:
@nperez0111 this does cause unnecessary renders (try adding a log in BlockNoteAIAU)

We could solve it by lifting blockId and status to the root, but then we lose Typescript help (see comment on line 26 of file). any insights?